File: src/claude_setup/domain/protocol_mapping.py
from __future__ import annotations

from pathlib import Path
from typing import Dict, List

from claude_setup.models import ProjectConfig

INTERFACE_PROTOCOL_MAP: Dict[str, List[str]] = {
    "rest": ["rest"],
    "grpc": ["grpc"],
    "graphql": ["graphql"],
    "websocket": ["websocket"],
    "event-consumer": ["event-driven", "messaging"],
    "event-producer": ["event-driven", "messaging"],
    "cli": [],
}

EVENT_PREFIX = "event-"
EVENT_DRIVEN_PROTOCOL = "event-driven"


def derive_protocols(config: ProjectConfig) -> List[str]:
    """Map interface types to protocol directory names.

    Returns deduplicated, sorted list of protocol names.
    """
    protocols: List[str] = []
    for iface in config.interfaces:
        mapped = INTERFACE_PROTOCOL_MAP.get(iface.type)
        if mapped is not None:
            protocols.extend(mapped)
        elif iface.type.startswith(EVENT_PREFIX):
            protocols.append(EVENT_DRIVEN_PROTOCOL)
    return sorted(set(protocols))


def derive_protocol_files(
    resources_dir: Path,
    protocol_names: List[str],
    config: ProjectConfig,
) -> Dict[str, List[Path]]:
    """List .md files per protocol directory.

    Applies broker-specific filtering for messaging protocol.
    Skips missing directories without error.
    """
    protocols_root = resources_dir / "protocols"
    result: Dict[str, List[Path]] = {}
    for protocol in protocol_names:
        protocol_dir = protocols_root / protocol
        if not protocol_dir.is_dir():
            continue
        if protocol == "messaging":
            files = _select_messaging_files(
                protocol_dir, config,
            )
        else:
            files = sorted(protocol_dir.glob("*.md"))
        if files:
            result[protocol] = files
    return result


def _select_messaging_files(
    messaging_dir: Path,
    config: ProjectConfig,
) -> List[Path]:
    """Select broker-specific or all messaging files."""
    broker = _extract_broker(config)
    if broker:
        specific = messaging_dir / f"{broker}.md"
        if specific.is_file():
            return [specific]
    return sorted(messaging_dir.glob("*.md"))


def _extract_broker(config: ProjectConfig) -> str:
    """Extract broker name from first interface that has one."""
    for iface in config.interfaces:
        if iface.broker:
            return iface.broker
    return ""

File: src/domain/protocol-mapping.ts
/**
 * Protocol derivation from interface types.
 *
 * Migrated from Python `domain/protocol_mapping.py`.
 */

import type { ProjectConfig } from "../models.js";

import { readdirSync, statSync } from "node:fs";
import { join } from "node:path";

/** Interface type to protocol directory names. */
export const INTERFACE_PROTOCOL_MAP: Readonly<Record<string, readonly string[]>> = {
  "rest": ["rest"],
  "grpc": ["grpc"],
  "graphql": ["graphql"],
  "websocket": ["websocket"],
  "event-consumer": ["event-driven", "messaging"],
  "event-producer": ["event-driven", "messaging"],
  "cli": [],
};

export const EVENT_PREFIX = "event-";
export const EVENT_DRIVEN_PROTOCOL = "event-driven";

/**
 * Map interface types to protocol directory names.
 *
 * Returns deduplicated, sorted list of protocol names.
 */
export function deriveProtocols(config: ProjectConfig): string[] {
  const protocols: string[] = [];
  for (const iface of config.interfaces) {
    const mapped = INTERFACE_PROTOCOL_MAP[iface.type];
    if (mapped !== undefined) {
      protocols.push(...mapped);
    } else if (iface.type.startsWith(EVENT_PREFIX)) {
      protocols.push(EVENT_DRIVEN_PROTOCOL);
    }
  }
  return [...new Set(protocols)].sort();
}

/**
 * List .md files per protocol directory.
 *
 * Applies broker-specific filtering for messaging protocol.
 * Skips missing directories without error.
 */
export function deriveProtocolFiles(
  resourcesDir: string,
  protocolNames: readonly string[],
  config: ProjectConfig,
): Record<string, string[]> {
  const protocolsRoot = join(resourcesDir, "protocols");
  const result: Record<string, string[]> = {};
  for (const protocol of protocolNames) {
    const protocolDir = join(protocolsRoot, protocol);
    try {
      if (!statSync(protocolDir).isDirectory()) {
        continue;
      }
    } catch {
      continue;
    }
    let files: string[];
    if (protocol === "messaging") {
      files = selectMessagingFiles(protocolDir, config);
    } else {
      files = readdirSync(protocolDir)
        .filter((f) => f.endsWith(".md"))
        .sort()
        .map((f) => join(protocolDir, f));
    }
    if (files.length > 0) {
      result[protocol] = files;
    }
  }
  return result;
}

/** Select broker-specific or all messaging files. */
export function selectMessagingFiles(
  messagingDir: string,
  config: ProjectConfig,
): string[] {
  const broker = extractBroker(config);
  if (broker) {
    const specific = join(messagingDir, `${broker}.md`);
    try {
      if (statSync(specific).isFile()) {
        return [specific];
      }
    } catch {
      // fall through to glob all
    }
  }
  return readdirSync(messagingDir)
    .filter((f) => f.endsWith(".md"))
    .sort()
    .map((f) => join(messagingDir, f));
}

/** Extract broker name from first interface that has one. */
export function extractBroker(config: ProjectConfig): string {
  for (const iface of config.interfaces) {
    if (iface.broker) {
      return iface.broker;
    }
  }
  return "";
}
